### PR TITLE
[1.20.5] Network API Refactor

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
This PR is the rebased and updated version of https://github.com/neoforged/NeoForge/pull/679, including all development from https://github.com/neoforged/Kits/pull/3.  There were very few rejects as compared to the Kits branch, so stringent review should not be necessary except for the patch files. 

For those who have not yet seen this PR, the changes are described below.

### Motivation
The current form of `IPayloadRegistrar` makes it difficult to register single-sided handlers (the most common type), as well as always executing the supplied handler on the network thread (when most handlers need to execute on the main thread).

### Changes
`IPayloadRegistrar` now has the `executesOn` method, which controls the default execution thread of supplied handlers.  The default thread is the main thread, but can be changed back to the network thread if desired.
When using the main thread, it is no longer necessary to use `enqueueWork` from within your payload handler.

The methods `IPayloadRegistrar#play` and`IPayloadRegistrar#configuration` were replaced with the following set of methods, respectively:
* `IPayloadRegistrar#playToClient`, `IPayloadRegistrar#playToServer`, `IPayloadRegistrar#playBidirectional`
* `IPayloadRegistrar#configurationToClient`, `IPayloadRegistrar#configurationToServer`, `IPayloadRegistrar#configurationBidirectional`

The method `IPayloadRegistrar#common` was removed.

Additonal complex payload handler registration can be done through `RegisterPayloadHandlersEvent#register`.

### Motivation
The payload context was previously composed of three (four for config) interfaces, each of which had one implementation per side (a total of 12 classes), on top of that, there were subclasses for each protocol, leading to what was effectively a maze of interfaces.  

Since these interfaces always have one concrete implementation (per-side), and were always constructed at the same time (using the same arguments, the packet listener), it is more reasonable to flatten them into the `IPayloadContext` object, and create two sided implementations of the context, usable by both protocols.

### Changes
The following classes have been removed: 
* `IReplyHandler`
  * `ClientReplyHandler`
  * `ServerReplyHandler`
* `IPacketHandler`
  * `ServerPacketHandler`
  * `ClientPacketHandler`
* `ISynchronizedWorkHandler`
  * `EventLoopSynchronizedWorkHandler`
* `ITaskCompletedHandler`,
* `ConfigurationPayloadContext`
* `IConfigurationPayloadHandler`
* `PlayPayloadContext`
* `IPlayPayloadHandler`

The following methods were added to `IPayloadContext`:
* `IPayloadContext#reply` - replacing `IReplyHandler#send`
* `IPayloadContext#disconnect` - replacing both `IReplyHandler#disconnect` and `IPacketHandler#disconnect`
* `IPayloadContext#handle` - replacing `IPacketHandler#handle` (including overloads)
* `IPayloadContext#enqueueWork` - replacing `ISynchronizedWorkHandler#submitAsync` (including overloads) with implicit exception handling
* `IPayloadContext#finishCurrentTask` - replacing `ITaskCompletedHandler#onTaskCompleted` 

All prior uses of `PlayPayloadContext` and `ConfigurationPayloadContext` have been loosened to `IPayloadContext`.

### Benefits

Flattening reduces indirection (consumers of the context can perform their actions immediately with the context object) and deduplicates a sizeable amount of code that was a result of the composition.
Additionally, we can reuse more code instead of fragmenting things on a per-protocol basis.

### Drawbacks
The method `IPayloadContext#finishCurrentTask` is now always visible, instead of only being visible to configuration-protocol payloads.  However, this is acceptable, as the benefit gained from not having to segment the paths is substantial.  We can make a reasonable assumption that users can read (we did so previously, because the method would throw on the client, even in configuration-protocol payloads).

### Motivation
All data within `NetworkRegistry` was static (including the existing non-static fields, due to the instantiation of `NetworkRegistry#INSTANCE`), but accessing this data required going through `NetworkRegistry.getInstance` which was a fairly pointless step.

### Changes
All fields and methods in `NetworkRegistry` were made static.  `NetworkRegistry#INSTANCE` and `NetworkRegistry#getInstance()` were removed.

## Rename of `isConnected` to `hasChannel`
See #459 

### Motivation
The event name is ill-formed for its purpose (which is to register additional configuration task objects to be run during the configuration phase).

### Changes
`OnGameConfigurationEvent` was renamed to `RegisterConfigurationTasksEvent`, which better reflects its purpose. For better compatibility, the event was also loosened to accept `ConfigurationTask` instances instead of only `ICustomConfigurationTask`.

### Motivation
There were many code paths within NetworkRegistry were either exact copies or near-exact copies of one another. By adjusting how we store certain data, we can centralize these code paths and not need multiple copies of the same thing.
### Changes
Code paths that were separated for play and configuration phases have been merged and now access the backing protocol-keyed `PAYLOAD_REGISTRATIONS` registry. Many data structures that also stored objects for both play and configuration phases reflect a similar change.
### Benefits
This allows us to automatically determine the correct map based on the protocol, which is already embedded in the `PacketListener`, instead of needing to check the type of the listener and redirect to the correct map.  
Additionally, if Mojang ever decides to support custom payloads on other protocols, or adds additional protocols which support them, we will not have to perform a major refactor of our data structures.

### CODE REVIEW

**Comment by Shadows-of-Fire** [comment]
If anyone wants to verify externally, the following files (from 24w09a) can be used to check a neo/fabric connection via ad-hoc payloads: https://gist.github.com/Shadows-of-Fire/f21d5ec5e92866233740e6fe47c61043

File: src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlerEvent.java [hunk 1]
```diff
@@ -5,31 +5,70 @@
...
+ * There are a few base rules for custom payloads:
+ * <ul>
+ * <li>Payloads should not write their ids within their {@link StreamCodec}, as this will be done automatically.</li>
+ * <li>Payloads must be registered before they can be sent over the network.</li>
+ * <li>Payloads may only be sent over the {@link ConnectionProtocol} they were registered for.</li>
+ * <li>By default, {@link IPayloadRegistrar} will wrap handlers so they execute on the main thread instead of the network thread.</li>
+ * <li>Payloads are always serialized, including for {@link Connection#isMemoryConnection() memory connections} (singleplayer and LAN hosts).</li>
+ * <ul>
  */
 public class RegisterPayloadHandlerEvent extends Event implements IModBusEvent {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlerEvent.java]: Any chance that we can rename this to `RegisterPayloadHandlersEvent`. (Note: Plural handler**s**).

File: src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java [hunk 2]
```diff
@@ -45,88 +42,74 @@
 
 @ApiStatus.Internal
 public class ClientPayloadHandler {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java]: make it `final`

File: src/main/java/net/neoforged/neoforge/network/handlers/ServerPayloadHandler.java [hunk 3]
```diff
@@ -6,21 +6,15 @@
 package net.neoforged.neoforge.network.handlers;
 
 import net.neoforged.neoforge.network.configuration.SyncRegistries;
-import net.neoforged.neoforge.network.handling.ConfigurationPayloadContext;
+import net.neoforged.neoforge.network.handling.IPayloadContext;
 import net.neoforged.neoforge.network.payload.FrozenRegistrySyncCompletedPayload;
 import org.jetbrains.annotations.ApiStatus;
 
 @ApiStatus.Internal
 public class ServerPayloadHandler {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/handlers/ServerPayloadHandler.java]: make it `final`

File: src/main/java/net/neoforged/neoforge/network/handling/DirectionalPayloadHandler.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Helper class that merges two unidirectional handlers into a single bidirectional handler.
+ */
+public record DirectionalPayloadHandler<T extends CustomPacketPayload>(IPayloadHandler<T> clientSide, IPayloadHandler<T> serverSide) implements IPayloadHandler<T> {
+    @Override
+    public void handle(T payload, IPayloadContext context) {
+        if (context.flow().isClientbound()) {
+            clientSide.handle(payload, context);
+
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/handling/DirectionalPayloadHandler.java]: nit: ugly whitespace

File: src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java [hunk 5]
```diff
@@ -6,32 +6,77 @@
...
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.network.ConfigurationTask;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.level.Level;
 
 /**
- * Defines a phase-less payload context that is passed to a handler for a payload that arrives during the connection.
+ * Common context interface for payload handlers.
  */
 public interface IPayloadContext {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java:27]: Should we have a getter for the `PacketListener` too? I think it could be useful. (Unfortunately we cannot make the typing more precise).

File: src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java [hunk 6]
```diff
@@ -5,190 +5,101 @@
...
-     * @param handler The handler for the payload.
-     * @return The registrar.
+     * When the handling thread is set to {@link HandlerThread#MAIN}, all registered handlers will be wrapped in {@link MainThreadPayloadHandler}.
+     * <p>
+     * The initial handling thread is {@link HandlerThread#MAIN}.
+     * 
+     * @param thread The default handling thread.
+     * @return A new registrar, which will map payload handlers to the desired thread.
      */
-    <T extends CustomPacketPayload> IPayloadRegistrar common(CustomPacketPayload.Type<T> id, StreamCodec<? super FriendlyByteBuf, T> reader, Consumer<IDirectionAwarePayloadHandlerBuilder<T, IPayloadHandler<T>>> handler);
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java:169]: Why did you remove the helpers to register common packets? They should be kept to match vanilla's patterns.

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java [hunk 7]
```diff
@@ -16,6 +18,9 @@
  * @param chosenVersion The chosen version, if any.
  */
 @ApiStatus.Internal
-public record NetworkChannel(
-        ResourceLocation id,
-        Optional<String> chosenVersion) {}
+public record NetworkChannel(ResourceLocation id, String chosenVersion) {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java:21]: where did the optional go? are versions mandatory now?

File: src/main/java/net/neoforged/neoforge/network/connection/ChannelAwareFriendlyByteBuf.java [hunk 8]
```diff
@@ -1,35 +0,0 @@
...
-package net.neoforged.neoforge.network.connection;
-
-import io.netty.buffer.ByteBuf;
-import io.netty.channel.ChannelHandlerContext;
-import net.minecraft.network.FriendlyByteBuf;
-import org.jetbrains.annotations.ApiStatus;
-import org.jetbrains.annotations.Nullable;
-
-@ApiStatus.Internal
-public final class ChannelAwareFriendlyByteBuf extends FriendlyByteBuf {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/connection/ChannelAwareFriendlyByteBuf.java:15]: Is there no way to the retrieve the `ChannelHandlerContext` from an FBB anymore? I have an upcoming use case for that actually. (Serializing custom ingredients as stack lists to vanilla clients by checking the connection attributes in the StreamCodec#encode).

**Review by Technici4n — CHANGES_REQUESTED** [review]
There are conflicts again (sorry!). The PR looks very good.

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java [hunk 9]
```diff
@@ -94,479 +85,273 @@ public class NetworkRegistry {
...
         if (setup)
             throw new IllegalStateException("The network registry can only be setup once.");
 
+        ModLoader.get().postEventWithReturn(new RegisterPayloadHandlerEvent(new NetworkRegistry()));
+
         setup = true;
+    }
+
+    /**
+     * This instance method is used instead of making an interface and passing that through to the event.
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java]: Cursed but OK. :P

File: src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlerEvent.java [hunk 10]
```diff
@@ -5,31 +5,70 @@
...
+     * @param handler   The handler for the payload. This handler should expect to receive the payload on all declared protocols and flows. It will be executed on the network thread.
+     * @param protocols The protocols this payload supports being sent over. Only {@link ConnectionProtocol#CONFIGURATION configuration} and {@link ConnectionProtocol#PLAY play} are supported.
+     * @param flow      The flow of this payload. Specify {@link Optional#empty()} to support sending in both directions.
+     * @param version   The version of the payload. Increase the payload version if the codec logic or handler logic changes. Neo-Neo connections with mismatched versions are denied.
+     * @param optional  If the payload is optional. Any connection with missing non-optional payloads is denied.
+     * @throws UnsupportedOperationException if the registration fails for any reason.
      */
-    public IPayloadRegistrar registrar(String namespace) {
-        return registrarFactory.apply(namespace);
+    public <T extends CustomPacketPayload, B extends FriendlyByteBuf> void register(CustomPacketPayload.Type<T> type, StreamCodec<? super B, T> codec, IPayloadHandler<T> handler,
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlerEvent.java]: I suggest we remove it and let all registration happen through the registrar. Then we can also remove the `NetworkRegistry` field and make `NetworkRegistry#register` static and package-private.

File: src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java [hunk 11]
```diff
@@ -5,190 +5,101 @@
...
- * Note the reader passed to any of the registration methods in this interface is invoked only if the payload is actually transferred over a connection which
- * is not marked as {@link Connection#isMemoryConnection()}. This is important for single-player of lan-opened worlds since there the writer and reader are
- * not invoked. That is because the payload is not actually transferred over the network, but only passed around in memory.
- * </p>
+ * Builder-style helper for registering {@link CustomPacketPayload}s, used for modded networking.
+ * 
+ * @see Class-level documentation on {@link RegisterPayloadHandlerEvent} for baseline rules.
+ * @see Method-level documentation on {@link RegisterPayloadHandlerEvent#register} for parameter information.
  */
 public interface IPayloadRegistrar {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java]: Should be inlined into `PayloadRegistrar`.

File: src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java (hunk 2 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/network/handlers/ClientPayloadHandler.java]: I don't think there's a point to bothering with `final` and `private` constructors for static-method classes, but eh, sure

File: src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java (hunk 5 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java:27]: I don't think there's any functionality that can be accomplished with the packet listener.  If needed, it can be retrieved from the player (the subclasses have public access to the strongly-typed listeners).

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java (hunk 7 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java:21]: Yes, though we have noted it will not make people use meaningful versions, there isn't any distinction between a useless version and an `Optional.empty()`, and the code flow is easier without having to deal with the `Optional`.

File: src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java (hunk 5 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java:27]: It can't be retrieved from the player in configuration contexts though.

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java (hunk 7 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java:21]: Potentially we can add a `registrar` method that defaults to a `""` version.

File: src/main/java/net/neoforged/neoforge/network/connection/ChannelAwareFriendlyByteBuf.java (hunk 8 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/network/connection/ChannelAwareFriendlyByteBuf.java:15]: Fine to remove for now - will add it back if necessary.

File: src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java (hunk 5 above)
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java:27]: Should add `@ApiStatus.NonExtendable`.

File: src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlersEvent.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Creates a new {@link PayloadRegistrar}, a utility for registering payloads using a builder-style format.
+     * 
+     * @param version The network version. May not be empty
+     */
+    public PayloadRegistrar registrar(String version) {
+        return new PayloadRegistrar(version);
+    }
+
+    public static enum HandlerThread {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlersEvent.java]: Would this be a better fit inside the `PayloadRegistrar` rather than the event? Or potentially as a top-level type?

File: src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java (hunk 5 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/network/handling/IPayloadContext.java:27]: Resolved to add it later if necessary; at face value it does not provide any useful functionality.

File: src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlersEvent.java (hunk 12 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/network/event/RegisterPayloadHandlersEvent.java]: Not sure. I had already shuffled it from the registrar to the event, but it feels a bit homeless.  I was trying to trim the size of the registrar class a bit

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java (hunk 7 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/neoforged/neoforge/network/registration/NetworkChannel.java:21]: Electing to leave as-is; network versions are probably going to be less relevant in the long term as we roll out the handshake protocol.

File: patches/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java.patch [hunk 13]
```diff
@@ -42,6 +65,33 @@
          LOGGER.warn("Client disconnected with reason: {}", p_295485_.getString());
      }
  
+@@ -302,7 +_,7 @@
+                         if (ClientCommonPacketListenerImpl.this.serverData != null) {
+                             ClientCommonPacketListenerImpl.this.serverData.setResourcePackStatus(ServerData.ServerPackStatus.ENABLED);
+                         }
+-
```
Comment (XFactHD) [inline patches/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java.patch]: What's going on with these three hunks?

File: src/main/java/net/neoforged/neoforge/network/handling/IPayloadHandler.java [hunk 14]
```diff
@@ -15,10 +18,11 @@
...
     /**
-     * Invoked to handle the given payload in the given context.
-     *
-     * @param payload The payload.
-     * @param context The context.
+     * Handles the payload with the supplied context.
+     * <p>
+     * When using {@link PayloadRegistrar} to register payloads, the executing thread defaults to the main thread. See {@link PayloadRegistrar#executesOn}.
+     * <p>
+     * When using {@link RegisterPayloadHandlersEvent#register}, the executing thread is always {@link HandlerThread#NETWORK}.
```
Comment (XFactHD) [inline src/main/java/net/neoforged/neoforge/network/handling/IPayloadHandler.java]: The method this is referring to doesn't exist anymore

**Review by XFactHD — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java.patch (hunk 13 above)
Comment (Shadows-of-Fire) [reply patches/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java.patch]: Looks like artifacts introduced by the rebase

**Review by Technici4n — APPROVED** [review]

**Review by XFactHD — APPROVED** [review]
